apps/io_tester/io_tester.cc

```diff
@@ -657,6 +665,96 @@ class write_io_class_data : public io_class_data {
     }
 };
 
+class unlink_class_data : public class_data {
+private:
+    sstring _dir_path{};
+    uint64_t _file_id_to_remove{0u};
+
+public:
+    unlink_class_data(job_config cfg) : class_data(std::move(cfg)) {
+        if (!_config.files_count.has_value()) {
+            throw std::runtime_error("request_type::unlink requires specifying 'files_count'");
+        }
+    }
+
+    future<> do_start(sstring path, directory_entry_type type) override {
+        if (type == directory_entry_type::directory) {
+            return do_start_on_directory(path);
+        }
+        throw std::runtime_error(format("Unsupported storage. {} should be directory", path));
+    }
+
+    future<size_t> issue_request(char *buf, io_intent* intent) override {
+        if (all_files_removed()) {
+            return make_ready_future<size_t>(1u);
+        }
+
+        const auto fname = get_filename(_file_id_to_remove);
+        ++_file_id_to_remove;
+
+        return remove_file(fname).then([]{
+            return make_ready_future<size_t>(1u);
+        });
+    }
+
+    void emit_results(YAML::Emitter& out) override {
+        out << YAML::Key << "average latency" << YAML::Value << average_latency();
+    }
+
+private:
+    uint64_t files_count() const {
+        return *_config.files_count;
+    }
+
+    bool all_files_removed() const {
+        return files_count() <= _file_id_to_remove;
+    }
+
+    sstring get_filename(uint64_t file_id) const {
+        return format("{}/test-{}-shard-{:d}-file-{}", _dir_path, name(), this_shard_id(), file_id);
+    }
+
+    future<> do_start_on_directory(sstring path) {
+        _dir_path = std::move(path);
+
+        return seastar::parallel_for_each(boost::irange(UINT64_C(0), files_count()), [this] (uint64_t file_id) {
```

Comment:
This makes unlink work at its maximum ~~speed~~ rate. However, more interesting case is when rate is configured and unlink tries to unlink "one file at N <x>seconds"

Reply:
Unlink is called in `future<size_t> issue_request(char *buf, io_intent* intent)` in line 687. Your comment refers to `do_start_on_directory()` which is a function called during startup to prepare files. Or at least I thought that it is called only during startup :sweat_smile: 

I agree, that calling `unlink()` could be configurable. I will try to implement the proposed comment.

However, I do not understand this particular statement - here `seastar::parallel_for_each()` is used to create files during startup and to ensure that they are created and filled at the maximum rate. These files are later unlinked in `issue_request()`.

This function is not run during evaluation (or I mistaken the way how main() works). However, from the code I would assume that `start()` finishes before starting the evaluation.

```
std::cout << "Creating initial files..." << std::endl;
ctx.invoke_on_all([] (auto& c) {
    return c.start();
}).get();

std::cout << "Starting evaluation..." << std::endl;
ctx.invoke_on_all([] (auto& c) {
    return c.issue_requests();
}).get();
```

Reply:
@xemul: is this conversation resolved? `unlink_class_data` is derived from `class_data`, which allows it to use `rpm` and `think_time`.

According to `io-tester.md`:
- **rps**: the requests-per-second rate to apply to sending fibers. If unset, each fiber sends new request as soon as previous one completes
- **think_time**: how long to wait before submitting another request in this job once one finishes.

The manual tests show that these parameters enable users to configure `request_type::unlink` to be aligned with your comment.